### PR TITLE
Remove last remaining C references to shmem allocator

### PR DESCRIPTION
**Pull request by sporksmith**
This should unblock merging the Rust rewrite of the shmem allocator without having to write a C API for it.

### CODE REVIEW

File: src/lib/shim/src/syscall.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                    // SAFETY: file descriptor should be valid and open.
+                    let strace_fd = unsafe { BorrowedFd::borrow_raw(strace_fd) };
+                    let mut strace_file_writer = BorrowedFdWriter::new(strace_fd);
+                    strace_file_writer.write_str(buffer.as_str()).unwrap();
+                }
+
+                return rv;
+            }
+            ShimEventToShim::Syscall(syscall) => {
+                let res = unsafe { native_syscall(&syscall.syscall_args) };
```
Comment (robgjansen) [inline src/lib/shim/src/syscall.rs:144]: Maybe add a comment about the difference here vs SyscallDoNative, since they are both calling `native_syscall`?

**Review by robgjansen — APPROVED** [review]
Looks fine to me. Since there are a fair bit of `unsafe` blocks here, I think it might make sense to ask @stevenengler to review this too, but I'll let you decide.

File: src/lib/shim/src/lib.rs [hunk 2]
```diff
@@ -718,4 +718,34 @@ pub mod export {
         // SAFETY: blk should be of the correct type and outlive this thread.
         unsafe { tls_thread_shmem::set(&thread_blk_serialized) };
     }
+
+    #[no_mangle]
+    pub extern "C" fn _shim_ipc_wait_for_start_event() {
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/lib/shim/src/lib.rs:686]: done

File: src/lib/shim/src/syscall.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        // This thread is exiting. Arrange for its thread-local-storage and
+        // signal stack to be freed.
+        unsafe { bindings::shim_freeSignalStack() };
+        // SAFETY: We don't try to recover from panics.
+        // TODO: make shim fully no_std and install a panic handler that aborts.
+        // https://doc.rust-lang.org/nomicon/panic-handler.html
+        unsafe { crate::release_and_exit_current_thread(exit_status) };
+    } else {
+        let rv: i64;
+        unsafe {
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/lib/shim/src/syscall.rs:41]: done

File: src/lib/shim/src/syscall.rs (hunk 1 above)
Comment (sporksmith) [reply src/lib/shim/src/syscall.rs:150]: done

**Comment by sporksmith** [comment]
> Looks fine to me. Since there are a fair bit of unsafe blocks here, I think it might make sense to ask @stevenengler to review this too, but I'll let you decide.

I think I'll go ahead and merge it rather than letting this sit for a week. @stevenengler if you'd like to review when you're back  happy to revisit and make changes :)

**Pull request merged**